### PR TITLE
Mark stage argument in hooks as required

### LINKED ISSUE #14062: RFC: Make stage argument to setup hooks formally required
### Motivation

The stage argument in the model hooks is currently annotated as 

```py
    def setup(self, stage: Optional[str] = None) -> None:
        ...
```

The framework never passes None in, hence the annotation is misleading and raises questions. This came up as a [question on StackOverflow](https://stackoverflow.com/questions/73130005/when-is-stage-is-none-in-pytorch-lightning/73261116#73261116).

Why was it done like this?
Quote from @carmocca:

> Because in the initial design thousand years ago, users were supposed to be able to do
> 
> ```py
> datamodule = MyDataModule()
> datamodule.setup()
> ...
> ```
> 
> where stage=None as a default would signal that all stages should be setup
> instead of
> 
> ```py
> datamodule.setup("fit")
> datamodule.setup("validate")
> datamodule.setup("test")
> ```

This is however no longer the recommended way. The setup method has long become a proper hook.

### Pitch

Change it to:

```py
    def setup(self, stage: str) -> None:
        ...
```

Is this a breaking change?
No, because users who have code with the "old" default value would still be able to run their code the same way. 

### Alternatives

Keep as is, but document that the None value is never passed in. 

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

cc @borda @tchaton @justusschock @awaelchli @carmocca @ninginthecloud @rohitgr7

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #14062

The stage argument was already implicitly required. The type hint suggesting that stage can be None is misleading, as this can never happen within the framework. It could only happen if the user manually calls the hook with None.

A change log entry shouldn't be necessary, as there is no behavioral change. But I can add one if reviewers request it. 

### Does your PR introduce any breaking changes? If yes, please list them.

No. Users with existing code that have stage=None in their signature will continue to have the exact same behavior. Since the framework does no longer make use of the optional value, this was a breaking change that was considered a long time ago already.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @rohitgr7 @justusschock @awaelchli

### CODE REVIEW

**Review by rohitgr7 — APPROVED** [review]
now `_check_setup_method` can be removed from `configuration_validator` along with associated test.

also changelog :)

**Comment by awaelchli** [comment]
@rohitgr7 Why are you adding a breaking change label? The change is not at all meant to be one. The update is exclusively the type-hint.

**Comment by rohitgr7** [comment]
it's a breaking change for the cases in LM, LDM and Callback:
```py
def setup(self, stage):
    super().setup()
```
which doesn't fail on master but does on this PR. But this case is very specific, so I am not sure if we should add this label. Removing it but add it if needed.

**Comment by Felonious-Spellfire** [comment]
lgtm

File: docs/source-pytorch/advanced/training_tricks.rst [hunk 1]
```diff
@@ -326,7 +326,7 @@ The :class:`~pytorch_lightning.core.datamodule.LightningDataModule` class provid
         def prepare_data(self):
             MNIST(self.data_dir, download=True)
 
-        def setup(self, stage: Optional[str] = None):
+        def setup(self, stage: str):
```
Comment (carmocca) [inline docs/source-pytorch/advanced/training_tricks.rst:329]: from platform import python_branch


I still see some instances that might need an update

```python
src/lightning_app/cli/pl-app-template/core/state.py:45:    stage: Optional[str] = None
src/pytorch_lightning/trainer/connectors/data_connector.py:245:        self, dataloader: Any, shuffle: Optional[bool] = None, mode: Optional[RunningStage] = None
src/pytorch_lightning/trainer/connectors/data_connector.py:296:        self, dataloader: DataLoader, shuffle: bool, mode: Optional[RunningStage] = None
src/pytorch_lightning/trainer/connectors/data_connector.py:333:        mode: Optional[RunningStage] = None,
src/pytorch_lightning/trainer/data_loading.py:30:    def prepare_dataloader(self, dataloader: Any, shuffle: bool, mode: Optional[RunningStage] = None) -> Any:
src/pytorch_lightning/profilers/profiler.py:42:        self._stage: Optional[str] = None
src/pytorch_lightning/profilers/profiler.py:151:    def setup(self, stage: str, local_rank: Optional[int] = None, log_dir: Optional[str] = None) -> None:
src/pytorch_lightning/profilers/profiler.py:157:    def teardown(self, stage: Optional[str]) -> None:
src/pytorch_lightning/profilers/advanced.py:81:    def teardown(self, stage: Optional[str]) -> None:
src/pytorch_lightning/utilities/data.py:197:    dataloader: DataLoader, sampler: Union[Sampler, Iterable], mode: Optional[RunningStage] = None
src/pytorch_lightning/utilities/data.py:207:    mode: Optional[RunningStage] = None,
src/pytorch_lightning/utilities/data.py:307:    mode: Optional[RunningStage] = None,
src/pytorch_lightning/strategies/ipu.py:241:        self, dataloader: DataLoader, sampler: Union[Sampler, Iterable], mode: Optional[RunningStage] = None
```

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]
Seems like a breaking change...

File: docs/source-pytorch/advanced/training_tricks.rst (hunk 1 above)
Comment (Borda) [inline docs/source-pytorch/advanced/training_tricks.rst:329]: also change log is missing

**Comment by carmocca** [comment]
The top post describes why a CHANGELOG entry was not added and why there's no behavioral impact
